### PR TITLE
GH-35390: [Python] Consolidate some APIs in Table and RecordBatch

### LINKED ISSUE #35390: [Python] Consolidate some identical pyarrow Table and RecordBatch APIs
### Describe the enhancement requested

Table and RecordBatch share identical implementations of many APIs. Let's put this in their shared parent class.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by danepitkin**
### Rationale for this change

Table and RecordBatch share many similar/identical APIs. Let's start consolidating them into the parent class. This diff refactors:

`__dataframe__`
`__eq__`
`__getitem__`
`_ensure_integer_index`
`column`
`columns`
`field`
`sort_by`

### Are these changes tested?

Relying on existing test cases for this refactor.

### Are there any user-facing changes?

No
* Closes: #35390

### CODE REVIEW

**Comment by danepitkin** [comment]
Build error is unrelated. The error exists on main (and in other PRs).

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -1963,106 +2113,16 @@ cdef class RecordBatch(_Tabular):
...
-                return field_indices[0]
-        elif isinstance(i, int):
-            return i
-        else:
-            raise TypeError("Index must either be string or integer")
+        return self._schema
 
-    def column(self, i):
+    @property
+    def columns(self):
```
Comment (AlenkaF) [inline python/pyarrow/table.pxi:2119]: Could we consolidate `columns` also? It seems both could be using `_column` or `column`.

File: python/pyarrow/table.pxi [hunk 2]
```diff
@@ -1500,6 +1627,68 @@ cdef class _Tabular(_PandasConvertible):
...
+            each entry is a tuple with column name
+            and sorting order ("ascending" or "descending")
+        **kwargs : dict, optional
+            Additional sorting options.
+            As allowed by :class:`SortOptions`
+
+        Returns
+        -------
+        Table or RecordBatch
+            A new tabular object sorted according to the sort keys.
```
Comment (AlenkaF) [inline python/pyarrow/table.pxi:1681]: Could we move the example from the `Table` `sort_by` to here?

**Review by AlenkaF — COMMENTED** [review]
Looking great! Adding just two comments.

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (danepitkin) [reply python/pyarrow/table.pxi:2119]: Absolutely!

**Comment by danepitkin** [comment]
Updated the diff to address comments!

**Comment by AlenkaF** [comment]
Thanks! Will merge, the failure is not related and the connected tests on that build are passing 👍

**Pull request merged**